### PR TITLE
Post additional telemetry from project system components

### PR DESCRIPTION
**Pull request by swesonga**
The managed project system runs various operations to support scenarios such as intellisense, compilation in the IDE, etc. Two critical functions of are:

1. Requesting that NuGet run a restore operation on a project ("nominating a project for restore"). Without this the data we pass to the language service may be incomplete.
2. Passing project data to the language service. Without this, the language service will have no data (and thus no useful Intellisense).

This PR updates the components involved in these scenarios to make them submit telemetry indicating which operations have been started and completed. A few of the components for scenario (1) include the package restore data sources and progress tracker (e.g. PackageRestoreConfiguredInputDataSource, PackageRestoreUnconfiguredInputDataSource, PackageRestoreDataSource, and PackageRestoreProgressTracker.

The first three are parts of a data flow chain, with the first harvesting relevant data from the Configured projects and the third actually asking NuGet to perform the restore operation. The last handles reporting to the Operation Progress system.

For scenario (2), the relevant types are those associated with Roslyn's IWorkspaceProjectContext. These include classes like ApplyChangesToWorkspaceContext (responsible for actually calling the methods on that interface as appropriate, based on the project changes), and ApplyChangesToWorkspaceContext (which manages the lifetime of the IWorkspaceProjectContext, invokes the ApplyChangesToWorkspaceContext as necessary, and reports status back to the Operation Progress system).

This description is a paraphrasing of a write-up by @tmeschter.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/6880)

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs:5]:
```suggestion
    internal static class PackageRestoreOperationNames
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.cs
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.cs:68]: I think this would work to remove some `!`s:

```suggestion
                    _projectId = Strings.IsNullOrEmpty(fullPath) ? string.Empty : _languageServiceTelemetryService.HashValue(fullPath);
                }

                return _projectId;
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/ILanguageServiceTelemetryService.cs
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/ILanguageServiceTelemetryService.cs:8]: The implemetnation of this interface imports `UnconfiguredProject`, so this interface should declare itself as being in the `UnconfiguredProject` scope.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/LanguageServiceOperationNames.cs
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/LanguageServiceOperationNames.cs:5]:
```suggestion
    internal static class LanguageServiceOperationNames
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreConfiguredInputDataSource.cs
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreConfiguredInputDataSource.cs:67]: Personally I would remove the new event classes. They don't seem to add much value, and require a heap allocation.

Instead, I'd move their constructor arguments to the `Log*Event` method directly, making the call sites simpler too:

```suggestion
            _packageReferenceTelemetryService.LogPackageRestoreEvent(PackageRestoreOperationNames.PackageRestoreConfiguredInputDataSourceLinkedToExternalInput, ContainingProject.FullPath);
```

One reason for using a class is if you expect to add members over time, as that can then be done in a non-API-breaking fashion (i.e. get/set property). However these are internal types so that's not a consideration in this case.

I don't feel too strongly about this. Just something to consider.

Alternatively convert them to `record`s. But I still don't think we need the runtime overhead or the extra API space.

Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreConfiguredInputDataSource.cs:67]: Most calls to these `Log*Event` methods seem to be passing a value that's derived from either the `ConfiguredProject` or `UnconfiguredProject`. Consider whether you can move the `IPackageRestoreTelemetryService`/`ILanguageServiceTelemetryService` interfaces into one of those scopes and have their implementations import the relevant projects directly, so that callers don't need to pass them on every call.

**Review by drewnoakes — COMMENTED** [review]
Some minor feedback below.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/ILanguageServiceTelemetryService.cs
Comment (swesonga) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/ILanguageServiceTelemetryService.cs:8]: Which implementation class are you referring to? I don't see the UnconfiguredProject import.

---
In reply to: [566549849](https://github.com/dotnet/project-system/pull/6880#discussion_r566549849) [](ancestors = 566549849)

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreConfiguredInputDataSource.cs
Comment (swesonga) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreConfiguredInputDataSource.cs:67]: I think this is a good idea. This was actually how I started out but thought about the API-breaking possibilities. You're right that this shouldn't be an issue since the API is internal.

---
In reply to: [566551743](https://github.com/dotnet/project-system/pull/6880#discussion_r566551743) [](ancestors = 566551743)

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreTelemetryService.cs
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreTelemetryService.cs:32]: Nit:
```suggestion
                    ( TelemetryPropertyName.PackageRestoreProjectId, projectId),
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreTelemetryService.cs
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreTelemetryService.cs:50]: Nit:
```suggestion
                    ( TelemetryPropertyName.PackageRestoreProjectId, projectId),
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreTelemetryService.cs
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreTelemetryService.cs:25]: Nit: `fullPath` might be better named as `projectFullPath` since all the calls use a project's path. Also, the description of the parameter says it is a project path too.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreProgressTracker.PackageRestoreProgressTrackerInstance.cs
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreProgressTracker.PackageRestoreProgressTrackerInstance.cs:99]: Nit: Might be best to replace `ValueTuple` with named ValueTuple via the `( )` syntax as you run into them. Otherwise, you end up with the calls below like `value.Value.Item1` which is not easily read on first glance.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/ILanguageServiceTelemetryService.cs
Comment (swesonga) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/ILanguageServiceTelemetryService.cs:9]: I got your point from the comment on the PackageRestoreConfiguredInputDataSource. Investigating this change.

---
In reply to: [566994525](https://github.com/dotnet/project-system/pull/6880#discussion_r566994525) [](ancestors = 566994525,566549849)

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs:10]: Nit:
```suggestion
        internal const string PackageRestoreDataSourceLinkedToExternalInput = nameof(PackageRestoreDataSourceLinkedToExternalInput);
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs:11]: Nit:
```suggestion
        internal const string PackageRestoreDataSourceLoading = nameof(PackageRestoreDataSourceLoading);
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs:12]: Nit:
```suggestion
        internal const string PackageRestoreDataSourceUnloading = nameof(PackageRestoreDataSourceUnloading);
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs:13]: Nit:
```suggestion
        internal const string PackageRestoreProgressTrackerActivating = nameof(PackageRestoreProgressTrackerActivating);
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs:14]: Nit:
```suggestion
        internal const string PackageRestoreProgressTrackerDeactivating = nameof(PackageRestoreProgressTrackerDeactivating);
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs:15]: Nit:
```suggestion
        internal const string PackageRestoreProgressTrackerInstanceInitialized = nameof(PackageRestoreProgressTrackerInstanceInitialized);
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs:16]: Nit:
```suggestion
        internal const string PackageRestoreProgressTrackerRestoreCompleted = nameof(PackageRestoreProgressTrackerRestoreCompleted);
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreOperationNames.cs:17]: Nit:
```suggestion
        internal const string PackageRestoreUnconfiguredInputDataSourceLinkedToExternalInput = nameof(PackageRestoreUnconfiguredInputDataSourceLinkedToExternalInput);
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/ILanguageServiceTelemetryService.cs
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/ILanguageServiceTelemetryService.cs:10]: I don't know the convention here, but would `ILanguageServiceTelemetryEvents` be a more accurate name? I don't know if this interface itself represents a 'Service`.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/ILanguageServiceTelemetryService.cs
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/ILanguageServiceTelemetryService.cs:17]: Since this name is a bit generic, would something relating to PII make sense to put in the name? Like, `HashPiiValue`?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/IPackageRestoreTelemetryService.cs
Comment (MiYanni) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/IPackageRestoreTelemetryService.cs:16]: Note: Left the comment for `fullPath` at implementation (should have probably been here). 😜

File: tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/Mocks/ILanguageServiceTelemetryServiceFactory.cs
Comment (MiYanni) [inline tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/Mocks/ILanguageServiceTelemetryServiceFactory.cs:13]: Is there a reason we prefer this boilerplate pattern over simply calling `new Mock` at the callsite?

File: tests/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/Mocks/IPackageRestoreTelemetryServiceFactory.cs
Comment (MiYanni) [inline tests/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/Mocks/IPackageRestoreTelemetryServiceFactory.cs:13]: Same boilerplate here as above comment.

**Review by MiYanni — APPROVED** [review]

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/LanguageServiceTelemetryService.cs
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/LanguageServiceTelemetryService.cs:32]: For interface implementation methods, if the docs are identical to those of the base I'd just remove them. There's less chance that they fall out of sync then.. The IDE will make documentation from the base visible here.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/LanguageServiceTelemetryService.cs
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/LanguageServiceTelemetryService.cs:37]: This `false` will be boxed. Might be good to reuse a boxed value by assigning it to a field.

CPS has an internal `Boxed` class for such values. Maybe we create one in this repo too.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.cs
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.cs:69]: As I understand it, this value is only used for telemetry, but its name suggests something grander.

If the language service telemetry component was in the `ConfiguredProject` scope, it could import the `ConfiguredProject` via its constructor and cache this value itself, keeping this telemetry concept out of the language service code. It seems you could then remove `HashValue` from the interface as well.

Looking further, it seems that this usage is in the configured project, and a usage below is in the unconfigured project scope, so the approach I describe here would only work if you split that service in two. I'm less sold on the idea, but thought I'd leave this thought here in case it prompts something.

At the very least, consider renaming to something like `TelemetryId`.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreTelemetryService.cs
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/PackageRestoreTelemetryService.cs:25]: If the `UnconfiguredProject` was imported into the constructor, I think these `fullPath` parameters could be removed.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.cs
Comment (swesonga) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.cs]: I have split the interface into 2 in [https://github.com/dotnet/project-system/commit/d718c99e5c46290d4ee63327c83188346ece0b9f](https://github.com/dotnet/project-system/commit/d718c99e5c46290d4ee63327c83188346ece0b9f) so that unconfigured and configured scopes are handled separately since the most likely change that could be needed is reporting additional telemetry and this seems like the cleaner way to handle that.

---
In reply to: [567238918](https://github.com/dotnet/project-system/pull/6880#discussion_r567238918) [](ancestors = 567238918)

Comment (swesonga) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.cs]: Still working on a fix for the duplication of the ProjectTelemetryId property though.

---
In reply to: [567490682](https://github.com/dotnet/project-system/pull/6880#discussion_r567490682) [](ancestors = 567490682,567238918)

File: tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/Mocks/ILanguageServiceTelemetryServiceFactory.cs
Comment (swesonga) [reply tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/Mocks/ILanguageServiceTelemetryServiceFactory.cs]: I did this to be consistent with the other tests I looked at. My interpretation was that the goal of this usage of the factory pattern is to abstract the exact manner in which an implementation of the returned interface is created so that, for example, a switch from XUnit to Moq to something else doesn't need the individual tests to be modified. Perhaps other folks might have some thoughts about this?

---
In reply to: [567169220](https://github.com/dotnet/project-system/pull/6880#discussion_r567169220) [](ancestors = 567169220)

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/ILanguageServiceTelemetryService.cs
Comment (swesonga) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/ILanguageServiceTelemetryService.cs]: Out of curiosity, what do you consider a service? Would this apply to the existing ITelemetryService as well?

---
In reply to: [567168061](https://github.com/dotnet/project-system/pull/6880#discussion_r567168061) [](ancestors = 567168061)

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/ConfiguredProjectLanguageServiceTelemetryService.cs
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/ConfiguredProjectLanguageServiceTelemetryService.cs:27]: Should this just be a null check? It can be assigned `""` in this block. I wouldn't expect `project.FullPath` to become non-null at any point. Ditto in other service classes.

This chunk of code is repeated a few times in this PR. Perhaps you could push it into the `HashValue` method (under a slightly different name), making this property something like:

```c#
private string ProjectTelemetryId => _projectHash ??= _telemetryService.GetProjectId(_project?.UnconfiguredProject);
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/ConfiguredProjectLanguageServiceTelemetryService.cs
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/ConfiguredProjectLanguageServiceTelemetryService.cs:41]: nit: spaces after opening paren here and elsewhere.

**Review by drewnoakes — APPROVED** [review]

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/ConfiguredProjectLanguageServiceTelemetryService.cs
Comment (swesonga) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/ConfiguredProjectLanguageServiceTelemetryService.cs:41]: I assumed that any spacing not changed by Edit > Advanced > Format Document was fine. Will fix.

---
In reply to: [567515572](https://github.com/dotnet/project-system/pull/6880#discussion_r567515572) [](ancestors = 567515572)

Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/ConfiguredProjectLanguageServiceTelemetryService.cs:41]: That's fair. I guess either our code style rules are underspecified, or it's not possible to control this particular styling quirk. It's not a big deal, so don't block merging for this.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreProgressTracker.PackageRestoreProgressTrackerInstance.cs
Comment (swesonga) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreProgressTracker.PackageRestoreProgressTrackerInstance.cs:99]: makes sense. I will change that in a follow-up PR though.

---
In reply to: [567164943](https://github.com/dotnet/project-system/pull/6880#discussion_r567164943) [](ancestors = 567164943)

**Review by tmeschter — CHANGES_REQUESTED** [review]
What I would like to see before this is merged in is an explanation about why we're capturing this information, and how we're going to use each event. I want to make sure we understand why it is important to us, and that we've thought through how we're going to process and report the data. For example, how are we going to use the telemetry related to the workspace project context to figure out if there are problems passing data to the language service?

I think the doc comments for each specific `Log*Event` method would be a good place to capture this information.

**Comment by swesonga** [comment]
> What I would like to see before this is merged in is an explanation about why we're capturing this information, and how we're going to use each event. I want to make sure we understand why it is important to us, and that we've thought through how we're going to process and report the data. For example, how are we going to use the telemetry related to the workspace project context to figure out if there are problems passing data to the language service?
> 
> I think the doc comments for each specific `Log*Event` method would be a good place to capture this information.

@jjmew  has provided the same feedback. I'm writing this up.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/IConfiguredProjectLanguageServiceTelemetryService.cs
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/IConfiguredProjectLanguageServiceTelemetryService.cs:9]: `/**` has no special meaning in C# (unlike in Java). This could be a `///` style comment instead. A few other instances below too.

**Review by drewnoakes — APPROVED** [review]

**Comment by swesonga** [comment]
@tmeschter I've called out the key events in the IConfiguredProjectLanguageServiceTelemetryService and IConfiguredProjectPackageRestoreTelemetryService interfaces. I think using the events called out there is rudimentary but probably sufficient for a first iteration.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/IConfiguredProjectLanguageServiceTelemetryService.cs
Comment (swesonga) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/IConfiguredProjectLanguageServiceTelemetryService.cs:9]: wow, muscle memory - definitely didn't catch myself doing this :D

---
In reply to: [569849335](https://github.com/dotnet/project-system/pull/6880#discussion_r569849335) [](ancestors = 569849335)

**Review by tmeschter — APPROVED** [review]

**Pull request merged**
